src/lib/common/include/sol-power-supply.h

```diff
@@ -0,0 +1,360 @@
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <stdlib.h>
+
+#include <sol-vector.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @file
+ * @brief Power Supply API provides a way to get information about battery,
+ * UPS, AC or DC power supply.
+ *
+ * Some information only is provided by some kinds of power supplies and
+ * depends on platform and drivers.
+ */
+
+/**
+ * @defgroup PowerSupply Power Supply
+ *
+ * @{
+ */
+
+enum sol_power_supply_type {
+    SOL_POWER_SUPPLY_TYPE_UNKNOWN,
+    SOL_POWER_SUPPLY_TYPE_BATTERY,
+    SOL_POWER_SUPPLY_TYPE_UPS,
+    SOL_POWER_SUPPLY_TYPE_MAINS,
+    SOL_POWER_SUPPLY_TYPE_USB,
+    SOL_POWER_SUPPLY_TYPE_USB_DCP,
+    SOL_POWER_SUPPLY_TYPE_USB_CDP,
+    SOL_POWER_SUPPLY_TYPE_USB_ACA
+};
+
+enum sol_power_supply_status {
+    SOL_POWER_SUPPLY_STATUS_UNKNOWN,
+    SOL_POWER_SUPPLY_STATUS_CHARGING,
+    SOL_POWER_SUPPLY_STATUS_DISCHARGING,
+    SOL_POWER_SUPPLY_STATUS_NOT_CHARGING,
+    SOL_POWER_SUPPLY_STATUS_FULL
+};
+
+enum sol_power_supply_capacity_level {
+    SOL_POWER_SUPPLY_CAPACITY_LEVEL_UNKNOWN,
+    SOL_POWER_SUPPLY_CAPACITY_LEVEL_CRITICAL,
+    SOL_POWER_SUPPLY_CAPACITY_LEVEL_LOW,
+    SOL_POWER_SUPPLY_CAPACITY_LEVEL_NORMAL,
+    SOL_POWER_SUPPLY_CAPACITY_LEVEL_HIGH,
+    SOL_POWER_SUPPLY_CAPACITY_LEVEL_FULL
+};
+
+/**
+ * Retrieves a list of names of all power supplies found. Types may vary.
+ *
+ * @param list A vector that will be initialized and filled with all
+ * power supply names found. It's required to free names pointed by
+ * it and clear the vector after usage.
+ *
+ * @return On success, it returns @c 0. On error, a negative value is returned
+ * and properties won't be set (pointers won't be valid);
+ */
+int sol_power_supply_get_list(struct sol_ptr_vector *list);
+
+/**
+ * Retrieves a list of names of all power supplies found that match
+ * a specified type.
+ *
+ * @param list A vector that will be initialized and filled with all
+ * power supply names found. It's required to free names pointed by
+ * it and clear the vector after usage.
+ *
+ * @param type Type of power supply to be fetched.
+ *
+ * @return On success, it returns @c 0. On error, a negative value is returned
+ * and properties won't be set (pointers won't be valid);
+ */
+int sol_power_supply_get_list_by_type(struct sol_ptr_vector *list, enum sol_power_supply_type type);
+
+/**
+ * Get type of power supply (battery, USB, ...)
+ *
+ * @param name Name of power supply. A list of all board supplies names may be
+ * fetched with sol_power_supply_get_list().
+ *
+ * @param type Type of power supply will be set on this pointer on success.
+ *
+ * @return On success, it returns @c 0. On error, a negative value is returned
+ * and properties won't be set (pointers won't be valid);
+ */
+int sol_power_supply_get_type(const char *name, enum sol_power_supply_type *type);
```

Comment:
It may returns < 0 that are valid for some integers, like charge family, as stated on kernel documentation.
It was the reason I decided to go with this approach. Also it guarantees an uniform API regarding error checks, and I don't need to make hacks like this one you're suggesting, to return integers for enums and booleans.

What makes it "strange" or hard to use in your opnion? I don't get it.